open-ce/test_feedstock.py

```diff
@@ -0,0 +1,223 @@
+
+    def get_test_command(self, conda_env_file=None):
+        """"
+        Returns a string of the test command.
+
+        Args:
+            conda_env_file (str): The name of the original conda environment file.
+                                  This is only needed when create_env is True.
+        """
+        output = ""
+        if self.create_env:
+            output += "conda env create -f " + conda_env_file + " -n " + self.conda_env + "\n"
+            return output
+
+        if self.clean_env:
+            output += "conda env remove -y -n " + self.conda_env + "\n"
+            return output
+
+        output += "CONDA_BIN=$(dirname $(which conda))\n"
+        output += "source ${CONDA_BIN}/../etc/profile.d/conda.sh\n"
+        output += "conda activate " + self.conda_env + "\n"
+        output += self.bash_command + "\n"
+
+        return output
+
+    def run(self, conda_env_file):
+        """
+        Runs the test.
+
+        Creates a temporary bash file, and writes the contents to `get_test_command` into it.
+        Runs the generated bash file.
+        Removes the temporary bash file.
+
+        Args:
+            conda_env_file (str): The name of the original conda environment file.
+        """
+        print("Running: " + self.name)
+        # Create file containing bash commands
+        with tempfile.NamedTemporaryFile(mode='w+t', dir=os.getcwd(), delete=False) as temp:
+            temp.write("set -e\n")
+            temp.write(self.get_test_command(conda_env_file))
+            temp_file_name = temp.name
+
+        # Execute file
+        test_process = subprocess.run(["bash", temp_file_name], stdout=subprocess.PIPE, check=False,
+                                      stderr=subprocess.STDOUT, universal_newlines=True)
+
+        # Remove file containing bash commands
+        os.remove(temp_file_name)
+
+        result = TestResult(self.name, test_process.returncode, test_process.stdout)
+
+        if test_process.returncode != 0:
+            result.display_failed()
+
+        return result
+
+class TestResult():
+    """
+    Contains the results of running a test.
+
+    Args:
+        name (str): The name of the test that was run.
+        returncode (int): The return code of the test that was run.
+        output (str): The resuling output from running the test.
+    """
+    def __init__(self, name, returncode, output):
+        self.name = name
+        self.output = output
+        self.returncode = returncode
+
+    def display_failed(self):
+        """
+        Display the output from a failed test.
+        """
+        if self.failed():
+            print("--------------------------------")
+            print("Failed test: " + self.name)
+            print(self.output)
+            print("--------------------------------")
+
+    def failed(self):
+        """
+        Returns whether or not a test failed.
+        """
+        return self.returncode != 0
+
+def load_test_file(test_file):
```

Comment:
This looks general enough to be in `utils.py` perhaps.

Reply:
I think this will eventually become a little more complicated where it is using variants in the same manner as the Open-CE environment files. At that point we can probably combine the two.

Reply:
ok